### PR TITLE
Spawn initial handshakes in separated task

### LINKED ISSUE #2163: Security: Spawn each initial peer handshake in a separate task, Credit: Equilibrium
**Is your feature request related to a problem? Please describe.**

Zebra can hang if there are a small number of initial peers. (It eventually continues after 3*75 seconds, but that's a long time to wait.)

There might also be some potential deadlocks between the initial peer handshakes, the candidate set, and the peer set.

Reported by Niklas Long of Equilibrium.

**Describe the solution you'd like**

Spawn each initial peer handshake in a separate task, so they can make progress independently.

**Additional context**

This is the same fix as #1950, but for the initial peer handshakes.

This is a follow-up to #2154, which fixed part of this security issue.

### PR DESCRIPTION
**Pull request by oxarbitrage**
## Motivation

We want to spawn initial peer handshakes in separated tasks. Closes https://github.com/ZcashFoundation/zebra/issues/2163

## Solution

Make the needed changes to spawn this connections in new task.

## Review

Anyone can review, the code is short enough.

### Reviewer Checklist

  - [ ] Current tests pass
  - [ ] Zebra is synchronizing

### CODE REVIEW

File: zebra-network/src/peer_set/initialize.rs [hunk 1]
```diff
@@ -285,20 +287,23 @@ where
...
+            // Construct a connector future but do not drive it yet ...
+            let outbound_connector_future = outbound_connector
+                .clone()
+                .oneshot(req)
+                .map_err(move |e| (addr, e));
+
+            // ... instead, spawn a new task to handle this connector
+            tokio::spawn(async move {
+                let task = outbound_connector_future.await;
+                // Only spawn one outbound connector per `MIN_PEER_CONNECTION_INTERVAL`.
```
Comment (teor2345) [inline zebra-network/src/peer_set/initialize.rs]: Nit: let's explain the multiplication, using a similar comment to the previous code

```suggestion
                // Only spawn one outbound connector per `MIN_PEER_CONNECTION_INTERVAL`,
                // sleeping for an interval according to its index in the list.
```

File: zebra-network/src/peer_set/initialize.rs [hunk 2]
```diff
@@ -285,20 +287,23 @@ where
...
-                    .map_err(move |e| (addr, e))
-                    .await
-            }
+                task
+            })
         })
         .collect();
 
-    while let Some(handshake_result) = handshakes.next().await {
+    while let Some(Ok(handshake_result)) = handshakes.next().await {
```
Comment (teor2345) [inline zebra-network/src/peer_set/initialize.rs]: Error handling: Can we turn handshake task panics into panics in the calling task?

If we just check for `Ok`, we will stop connecting to peers if a handshake panics, and continue running Zebra. So we won't actually see the panics in the logs.

(I can do a similar fix for inbound connections.)

```suggestion
    while let Some(handshake_result) = handshakes.next().await {
        let handshake_result = handshake_result.expect("panic in initial peer handshake");
```

**Review by teor2345 — COMMENTED** [review]
Looks great!

The current code ignores panics in the spawned tasks, but otherwise this is looking pretty good.

File: zebra-network/src/peer_set/initialize.rs (hunk 1 above)
Comment (oxarbitrage) [reply zebra-network/src/peer_set/initialize.rs]: https://github.com/ZcashFoundation/zebra/pull/3189/commits/ec0099fe1a406713b63fd2c18c417cdacf55f737

File: zebra-network/src/peer_set/initialize.rs (hunk 2 above)
Comment (oxarbitrage) [reply zebra-network/src/peer_set/initialize.rs]: Thanks, changed to this at https://github.com/ZcashFoundation/zebra/pull/3189/commits/36f708c82875c1a1ddc0858eb284fd948ee74688

**Review by teor2345 — APPROVED** [review]
All done, thanks!

**Pull request merged**